File: shape_bit_field.go
package puyo2

import (
	"fmt"
	"math/bits"
	"os"
	"strings"
)

type ShapeBitField struct {
	Shapes             []*FieldBits
	originalShapes     []*FieldBits
	ChainOrderedShapes [][]*FieldBits
	keyString          *string
}

func NewShapeBitField() *ShapeBitField {
	shapeBitField := new(ShapeBitField)
	return shapeBitField
}

var letter2Num = map[rune]int{
	'1': 1,
	'2': 2,
	'3': 3,
	'4': 4,
	'5': 5,
	'6': 6,
	'7': 7,
	'8': 8,
	'9': 9,
	'a': 10,
	'b': 11,
	'c': 12,
	'd': 13,
	'e': 14,
	'f': 15,
	'g': 16,
	'h': 17,
	'i': 18,
	'j': 19,
	'k': 20,
}

var num2Letter = map[int]string{
	1:  "1",
	2:  "2",
	3:  "3",
	4:  "4",
	5:  "5",
	6:  "6",
	7:  "7",
	8:  "8",
	9:  "9",
	10: "a",
	11: "b",
	12: "c",
	13: "d",
	14: "e",
	15: "f",
	16: "g",
	17: "h",
	18: "i",
	19: "j",
	20: "k",
}

func NewShapeBitFieldWithFieldString(param string) *ShapeBitField {
	shapeBitField := new(ShapeBitField)

	max := 0
	for _, c := range param {
		if c == '.' {
			continue
		}
		n := letter2Num[c]
		if n > max {
			max = n
		}
	}
	shapes := make([]*FieldBits, max)
	for i := 0; i < max; i++ {
		shapes[i] = NewFieldBits()
	}
	for i, c := range param {
		if c == '.' {
			continue
		}
		n := letter2Num[c]
		x := i % 6
		y := 13 - i/6

		shapes[n-1].SetOnebit(x, y)
	}

	for _, shape := range shapes {
		shapeBitField.AddShape(shape)
	}

	return shapeBitField
}

func (sbf *ShapeBitField) AddShape(shape *FieldBits) {
	sbf.Shapes = append(sbf.Shapes, shape)
	sbf.originalShapes = append(sbf.originalShapes, shape.Clone())
}

func (sbf *ShapeBitField) Clone() *ShapeBitField {
	shapeBitField := new(ShapeBitField)
	shapeBitField.Shapes = make([]*FieldBits, len(sbf.Shapes))
	for i, shape := range sbf.Shapes {
		shapeBitField.Shapes[i] = shape.Clone()
	}
	shapeBitField.originalShapes = make([]*FieldBits, len(sbf.originalShapes))
	for i, originalShape := range sbf.originalShapes {
		shapeBitField.originalShapes[i] = originalShape.Clone()
	}
	return shapeBitField
}

func (sbf *ShapeBitField) Drop(fb *FieldBits) {
	for _, shape := range sbf.Shapes {
		r0 := Extract(shape.M[0], ^fb.M[0])
		r1 := Extract(shape.M[1], ^fb.M[1])
		var dropmask1 [2]uint64
		for x := 0; x < 6; x++ {
			idx := x >> 2
			vc := bits.OnesCount64(fb.ColBits(x))
			dropmask1[idx] |= bits.RotateLeft64((1<<vc)-1, 14-vc) << (x & 3 * 16)
		}
		shape.M[0] = Deposit(r0, ^dropmask1[0])
		shape.M[1] = Deposit(r1, ^dropmask1[1])
	}
}

func (sbf *ShapeBitField) findPossibilities(colors []Color, num int) []Color {
	if colors[num] != Empty {
		return []Color{colors[num]}
	}
	palet := map[Color]struct{}{}
	order := []Color{Red, Blue, Yellow, Green}
	for _, c := range order {
		palet[c] = struct{}{}
	}
	for i, shape := range sbf.Shapes {
		if i == num {
			continue
		}
		ex := sbf.Shapes[num].Expand(shape)
		if ex.IsEmpty() == false && ex.FindVanishingBits().IsEmpty() == false {
			delete(palet, colors[i])
		}
	}
	possibilities := []Color{}
	for _, c := range order {
		if _, ok := palet[c]; ok {
			possibilities = append(possibilities, c)
		}
	}
	return possibilities
}

func (sbf *ShapeBitField) findColor(targets []Color, idx int) *BitField {
	if idx == sbf.ShapeCount() {
		sbfArray := sbf.ToChainShapesUInt64Array()
		overall := sbf.OverallShape()

		bf := NewBitField()
		for x := 0; x < 6; x++ {
			for y := 0; y < 14; y++ {
				if overall.Onebit(x, y) > 0 {
					for i, shape := range sbf.Shapes {
						if shape.Onebit(x, y) > 0 {
							bf.SetColor(targets[i], x, y)
							break
						}
					}
				}
			}
		}
		cbf := bf.Clone()
		result := cbf.Simulate()
		if result.Chains == sbf.ShapeCount() {
			sameChain := true
			for i, array := range bf.ToChainShapesUInt64Array() {
				if sbfArray[i][0] != array[0] || sbfArray[i][1] != array[1] {
					sameChain = false
					break
				}
			}
			if sameChain {
				return bf
			} else {
				fmt.Fprintf(os.Stderr, "Whats!!!!!!!!! %s", bf.MattulwanEditorParam())
				return nil
			}
		}
		return nil
	}

	colors := sbf.findPossibilities(targets, idx)
	for _, c := range colors {
		newTargets := make([]Color, len(targets))
		copy(newTargets, targets)
		newTargets[idx] = c
		bf := sbf.findColor(newTargets, idx+1)
		if bf != nil {
			return bf
		}
	}
	return nil
}

func (sbf *ShapeBitField) FillChainableColor() *BitField {
	targets := make([]Color, sbf.ShapeCount())
	targets[0] = Red
	return sbf.findColor(targets, 0)
}

func (sbf *ShapeBitField) FindVanishingFieldBitsNum() []int {
	vfbn := []int{}
	for i, shape := range sbf.Shapes {
		vs := shape.MaskField12().FindVanishingBits()
		if !vs.IsEmpty() {
			vfbn = append(vfbn, i)
		}
	}
	return vfbn
}

func (sbf *ShapeBitField) IsEmpty() bool {
	for _, shape := range sbf.Shapes {
		if shape.IsEmpty() == false {
			return false
		}
	}
	return true
}

func (sbf *ShapeBitField) OriginalOverallShape() *FieldBits {
	fb := NewFieldBits()
	for _, shape := range sbf.originalShapes {
		fb = fb.Or(shape)
	}
	return fb
}

func (sbf *ShapeBitField) OverallShape() *FieldBits {
	fb := NewFieldBits()
	for _, shape := range sbf.Shapes {
		fb = fb.Or(shape)
	}
	return fb
}

func (sbf *ShapeBitField) ShapeCount() int {
	return len(sbf.Shapes)
}

func (sbf *ShapeBitField) ShapeNum(x int, y int) int {
	for i, shape := range sbf.Shapes {
		if shape.Onebit(x, y) > 0 {
			return i
		}
	}
	return -1
}

func (sbf *ShapeBitField) Simulate1() []int {
	vfbn := sbf.FindVanishingFieldBitsNum()
	v := NewFieldBits()
	for _, n := range vfbn {
		vf := sbf.Shapes[n]
		v = v.Or(vf.MaskField12().FindVanishingBits())
	}
	if v.IsEmpty() {
		return []int{}
	}
	sbf.Drop(v)

	chainShapes := []*FieldBits{}
	for _, n := range vfbn {
		chainShapes = append(chainShapes, sbf.originalShapes[n])
	}
	sbf.ChainOrderedShapes = append(sbf.ChainOrderedShapes, chainShapes)
	return vfbn
}

func (sbf *ShapeBitField) Simulate() *ShapeRensaResult {
	result := NewShapeRensaResult()
	for {
		vfbn := sbf.Simulate1()
		if len(vfbn) == 0 {
			break
		}
		result.AddChain()
	}
	result.SetShapeBitField(sbf)
	return result
}

func (sbf *ShapeBitField) shiftCol(insertShape *FieldBits, shiftTarget *FieldBits, colNum int) {
	for i, shape := range sbf.Shapes {
		and := shape.And(shiftTarget)

		shiftBits := 0
		colBits := and.ColBits(colNum)
		if colBits == 0 {
			continue
		}

		shiftBits = bits.OnesCount(uint(insertShape.ColBits(colNum)))
		shiftCol := colBits << shiftBits
		if colNum < 4 {
			shift := NewFieldBitsWithM([2]uint64{shiftCol, 0})
			unTarget := NewFieldBitsWithM([2]uint64{^(shiftTarget.M[0] | 1<<(colNum*16)), 0})
			unTarget.M[0] = unTarget.ColBits(colNum)
			unTarget = unTarget.And(shape)
			switch colNum {
			case 0:
				shape.M[0] &= ^uint64(0xffff)
			case 1:
				shape.M[0] &= ^uint64(0xffff0000)
			case 2:
				shape.M[0] &= ^uint64(0xffff00000000)
			case 3:
				shape.M[0] &= ^uint64(0xffff000000000000)
			}

			shape = shape.Or(shift)
			shape = shape.Or(unTarget)
		} else {
			shift := NewFieldBitsWithM([2]uint64{0, shiftCol})
			unTarget := NewFieldBitsWithM([2]uint64{0, ^(shiftTarget.M[1] | 1<<((colNum-4)*16))})
			unTarget.M[1] = unTarget.ColBits(colNum)
			unTarget = unTarget.And(shape)
			switch colNum {
			case 4:
				shape.M[1] &= ^uint64(0xffff)
			case 5:
				shape.M[1] &= ^uint64(0xffff0000)
			}

			shape = shape.Or(shift)
			shape = shape.Or(unTarget)

		}
		sbf.Shapes[i] = shape
		sbf.originalShapes[i] = shape.Clone()
	}
}

func (sbf *ShapeBitField) InsertShape(fb *FieldBits) {
	overall := sbf.OverallShape()
	and := overall.And(fb)
	for x := 0; x < 6; x++ {
		col := and.ColBits(x)
		if col == 0 {
			continue
		}
		shiftTarget := fb.Clone()
		sb := x * 16
		if x > 3 {
			sb = (x - 4) * 16
		}
		for y := bits.Len64(col >> sb); y < 16; y++ {
			shiftTarget.SetOnebit(x, y)
		}
		sbf.shiftCol(fb, shiftTarget, x)
	}
	sbf.AddShape(fb)
}

func (sbf *ShapeBitField) Expand3PuyoShapes() *ShapeBitField {
	csbf := sbf.Clone()
	for i, shape := range csbf.Shapes {
		if shape.PopCount() == 3 {
			overall := csbf.OverallShape()
			overall.M[0] = ^overall.M[0]
			overall.M[1] = ^overall.M[1]
			overall = overall.MaskField13()
			expand := shape.Expand1(overall)
			csbf.Shapes[i] = shape.Or(expand)
		}
	}
	return csbf
}

func (sbf *ShapeBitField) KeyString() string {
	if sbf.keyString != nil {
		return *sbf.keyString
	}

	var b = strings.Builder{}
	for _, shape := range sbf.Shapes {
		fmt.Fprintf(&b, "_%x:%x", shape.M[0], shape.M[1])
	}
	ks := b.String()
	sbf.keyString = &ks
	return ks

}

func (sbf *ShapeBitField) FieldString() string {
	buf := []string{".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", "."}
	overall := sbf.OverallShape()
	for y := 13; y > 0; y-- {
		for x := 0; x < 6; x++ {
			if overall.Onebit(x, y) == 0 {
				continue
			}
			idx := (13-y)*6 + x
			for i, shape := range sbf.Shapes {
				if shape.Onebit(x, y) > 0 {
					buf[idx] = num2Letter[i+1]
					break
				}
			}
		}
	}
	return strings.Join(buf, "")
}

func (sbf *ShapeBitField) ChainOrderedFieldString() string {
	shapes := []*FieldBits{}
	for _, shape := range sbf.ChainOrderedShapes {
		shapes = append(shapes, shape[0])
	}
	for i, shape := range sbf.Shapes {
		if shape.IsEmpty() == false {
			shapes = append(shapes, sbf.originalShapes[i])
		}
	}
	var b strings.Builder
	for y := 13; y > 0; y-- {
		for x := 0; x < 6; x++ {
			e := true
			for i, shape := range shapes {
				if shape.Onebit(x, y) > 0 {
					fmt.Fprint(&b, num2Letter[i+1])
					e = false
					break
				}
			}
			if e {
				fmt.Fprint(&b, ".")
			}
		}
	}
	return b.String()
}

func (sbf *ShapeBitField) ShowDebug() {
	fmt.Print(sbf.ToString())
}

func (sbf *ShapeBitField) ToString() string {
	var b strings.Builder
	b.Grow(165)
	for y := 14; y > 0; y-- {
		fmt.Fprintf(&b, "%02d: ", y)
		for x := 0; x < 6; x++ {
			e := true
			for i, shape := range sbf.Shapes {
				if shape.Onebit(x, y) > 0 {
					fmt.Fprint(&b, num2Letter[i+1])
					e = false
					break
				}
			}
			if e {
				fmt.Fprint(&b, ".")
			}
		}
		fmt.Fprintln(&b)
	}
	return b.String()
}

func (sbf *ShapeBitField) ToChainShapes() []*FieldBits {
	csbf := sbf.Clone()
	shapes := make([]*FieldBits, 0)
	vfbn := csbf.FindVanishingFieldBitsNum()
	if len(vfbn) == 0 {
		return shapes
	}
	shapes = append(shapes, sbf.Shapes[vfbn[0]])
	for {
		csbf.Simulate1()
		vfbn := csbf.FindVanishingFieldBitsNum()
		if len(vfbn) != 0 {
			shapes = append(shapes, csbf.Shapes[vfbn[0]].Clone())
		} else {
			break
		}
	}
	return shapes
}

func (sbf *ShapeBitField) ToChainShapesUInt64Array() [][2]uint64 {
	shapes := sbf.ToChainShapes()
	array := make([][2]uint64, 0, len(shapes))
	for _, v := range shapes {
		array = append(array, v.ToIntArray())
	}
	return array
}

File: crates/puyo2/src/shape_bit_field.rs
use std::{fmt, io, path::Path};

use serde::{Deserialize, Serialize};

use crate::{
    BitField, Color, FieldBits, ShapeRensaResult,
    drop_compact::compact_lane_u16,
    render::{TILE_SIZE, copy_tile, load_sprite, new_canvas, overlay_tile, write_png},
};

const SHAPE_LABELS: [char; 20] = [
    '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
    'k',
];

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ShapeBitField {
    pub shapes: Vec<FieldBits>,
    pub original_shapes: Vec<FieldBits>,
    pub chain_ordered_shapes: Vec<Vec<FieldBits>>,
    pub key_string: Option<String>,
}

impl ShapeBitField {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn from_field_string(param: &str) -> Self {
        let mut shape_bit_field = Self::new();
        let max = param
            .chars()
            .filter_map(letter_to_num)
            .max()
            .unwrap_or_default();
        let mut shapes = vec![FieldBits::new(); max];

        for (index, ch) in param.chars().enumerate() {
            if ch == '.' {
                continue;
            }
            let n = letter_to_num(ch)
                .unwrap_or_else(|| panic!("unsupported shape label {ch:?} in field string"));
            let x = index % 6;
            let y = 13usize.saturating_sub(index / 6);
            shapes[n - 1].set_onebit(x, y);
        }

        for shape in shapes {
            shape_bit_field.add_shape(shape);
        }
        shape_bit_field
    }

    pub fn add_shape(&mut self, shape: FieldBits) {
        self.shapes.push(shape);
        self.original_shapes.push(shape);
        self.key_string = None;
    }

    pub fn drop(&mut self, field_bits: FieldBits) {
        let mut active = [(0usize, 0u32, 0u64, 0u16); 6];
        let mut active_len = 0usize;
        for x in 0..6 {
            let idx = x >> 2;
            let shift = ((x & 3) * 16) as u32;
            let vanished_lane = ((field_bits.m[idx] >> shift) & 0xffff) as u16;
            if vanished_lane == 0 {
                continue;
            }
            active[active_len] = (idx, shift, 0xffffu64 << shift, vanished_lane);
            active_len += 1;
        }
        for shape in &mut self.shapes {
            for &(idx, shift, lane_mask, vanished_lane) in &active[..active_len] {
                let lane = ((shape.m[idx] >> shift) & 0xffff) as u16;
                let compacted = compact_lane_u16(lane, vanished_lane) as u64;
                shape.m[idx] = (shape.m[idx] & !lane_mask) | (compacted << shift);
            }
        }
        self.key_string = None;
    }

    pub fn fill_chainable_color(&self) -> Option<BitField> {
        if self.shape_count() == 0 {
            return Some(BitField::new());
        }
        let mut targets = vec![Color::Empty; self.shape_count()];
        targets[0] = Color::Red;
        self.find_color(targets, 0)
    }

    pub fn find_vanishing_field_bits_num(&self) -> Vec<usize> {
        let mut vanishing = Vec::new();
        for (index, shape) in self.shapes.iter().enumerate() {
            let vanished = shape.mask_field12().find_vanishing_bits();
            if !vanished.is_empty() {
                vanishing.push(index);
            }
        }
        vanishing
    }

    pub fn is_empty(&self) -> bool {
        self.shapes.iter().all(FieldBits::is_empty)
    }

    pub fn original_overall_shape(&self) -> FieldBits {
        self.original_shapes
            .iter()
            .copied()
            .fold(FieldBits::new(), |acc, shape| acc.or(shape))
    }

    pub fn overall_shape(&self) -> FieldBits {
        self.shapes
            .iter()
            .copied()
            .fold(FieldBits::new(), |acc, shape| acc.or(shape))
    }

    pub fn shape_count(&self) -> usize {
        self.shapes.len()
    }

    pub fn shape_num(&self, x: usize, y: usize) -> isize {
        for (index, shape) in self.shapes.iter().enumerate() {
            if shape.onebit(x, y) > 0 {
                return index as isize;
            }
        }
        -1
    }

    pub fn simulate1(&mut self) -> Vec<usize> {
        let vanishing_nums = self.find_vanishing_field_bits_num();
        let mut vanished = FieldBits::new();
        for index in &vanishing_nums {
            vanished = vanished.or(self.shapes[*index].mask_field12().find_vanishing_bits());
        }
        if vanished.is_empty() {
            return Vec::new();
        }
        self.drop(vanished);

        let chain_shapes = vanishing_nums
            .iter()
            .map(|index| self.original_shapes[*index])
            .collect();
        self.chain_ordered_shapes.push(chain_shapes);
        vanishing_nums
    }

    pub fn simulate(&mut self) -> ShapeRensaResult {
        let mut result = ShapeRensaResult::new();
        loop {
            let vanished = self.simulate1();
            if vanished.is_empty() {
                break;
            }
            result.add_chain();
        }
        result.set_shape_bit_field(self.clone());
        result
    }

    pub fn insert_shape(&mut self, field_bits: FieldBits) {
        let overall = self.overall_shape();
        let and = overall.and(field_bits);
        for x in 0..6 {
            let col = and.col_bits(x);
            if col == 0 {
                continue;
            }
            let mut shift_target = field_bits;
            let shift_base = if x > 3 { (x - 4) * 16 } else { x * 16 };
            let start = 64usize.saturating_sub((col >> shift_base).leading_zeros() as usize);
            for y in start..16 {
                shift_target.set_onebit(x, y);
            }
            self.shift_col(field_bits, shift_target, x);
        }
        self.add_shape(field_bits);
    }

    pub fn expand_3_puyo_shapes(&self) -> Self {
        let mut cloned = self.clone();
        for index in 0..cloned.shapes.len() {
            if cloned.shapes[index].popcount() == 3 {
                let mut overall = cloned.overall_shape();
                overall.m[0] = !overall.m[0];
                overall.m[1] = !overall.m[1];
                overall = overall.mask_field13();
                let expanded = cloned.shapes[index].expand1(overall);
                cloned.shapes[index] = cloned.shapes[index].or(expanded);
            }
        }
        cloned.key_string = None;
        cloned
    }

    pub fn key_string(&mut self) -> &str {
        if self.key_string.is_none() {
            let mut output = String::new();
            for shape in &self.shapes {
                output.push('_');
                output.push_str(&format!("{:x}:{:x}", shape.m[0], shape.m[1]));
            }
            self.key_string = Some(output);
        }
        self.key_string.as_deref().unwrap()
    }

    pub fn field_string(&self) -> String {
        let mut buf = vec!['.'; 78];
        let overall = self.overall_shape();
        for y in (1..=13).rev() {
            for x in 0..6 {
                if overall.onebit(x, y) == 0 {
                    continue;
                }
                let idx = (13 - y) * 6 + x;
                for (shape_index, shape) in self.shapes.iter().enumerate() {
                    if shape.onebit(x, y) > 0 {
                        buf[idx] = num_to_letter(shape_index + 1);
                        break;
                    }
                }
            }
        }
        buf.into_iter().collect()
    }

    pub fn chain_ordered_field_string(&self) -> String {
        let mut ordered_shapes = Vec::new();
        for chain_shapes in &self.chain_ordered_shapes {
            if let Some(shape) = chain_shapes.first() {
                ordered_shapes.push(*shape);
            }
        }
        for (index, shape) in self.shapes.iter().enumerate() {
            if !shape.is_empty() {
                ordered_shapes.push(self.original_shapes[index]);
            }
        }

        let mut output = String::new();
        for y in (1..=13).rev() {
            for x in 0..6 {
                let mut empty = true;
                for (index, shape) in ordered_shapes.iter().enumerate() {
                    if shape.onebit(x, y) > 0 {
                        output.push(num_to_letter(index + 1));
                        empty = false;
                        break;
                    }
                }
                if empty {
                    output.push('.');
                }
            }
        }
        output
    }

    pub fn to_chain_shapes(&self) -> Vec<FieldBits> {
        let mut cloned = self.clone();
        let mut shapes = Vec::new();
        let vanishing_nums = cloned.find_vanishing_field_bits_num();
        if vanishing_nums.is_empty() {
            return shapes;
        }
        shapes.push(self.shapes[vanishing_nums[0]]);
        loop {
            cloned.simulate1();
            let vanishing_nums = cloned.find_vanishing_field_bits_num();
            if vanishing_nums.is_empty() {
                break;
            }
            shapes.push(cloned.shapes[vanishing_nums[0]]);
        }
        shapes
    }

    pub fn to_chain_shapes_u64_array(&self) -> Vec<[u64; 2]> {
        self.to_chain_shapes()
            .into_iter()
            .map(|shape| shape.to_int_array())
            .collect()
    }

    pub fn export_image<P: AsRef<Path>>(&self, name: P) -> io::Result<()> {
        let field = load_sprite("puyos.png")?;
        let puyo = load_sprite("puyos_shape.png")?;
        let mut out = new_canvas(8, 14);
        draw_field(&field, &mut out);

        for y in (1..=13).rev() {
            for x in 0..6 {
                for (index, shape) in self.shapes.iter().enumerate() {
                    if shape.onebit(x, y) > 0 {
                        draw_shape_puyo(shape, index as u32, x, y, &puyo, &mut out);
                    }
                }
            }
        }

        write_png(name.as_ref(), &out)
    }

    pub fn export_chain_image<P: AsRef<Path>>(&self, name: P) -> io::Result<()> {
        let field = load_sprite("puyos.png")?;
        let puyo = load_sprite("puyos_shape.png")?;
        let mut out = new_canvas(8, 14);
        draw_field(&field, &mut out);

        for y in (1..=13).rev() {
            for x in 0..6 {
                for (index, chain_shapes) in self.chain_ordered_shapes.iter().enumerate() {
                    for shape in chain_shapes {
                        if shape.onebit(x, y) > 0 {
                            draw_shape_puyo(shape, index as u32, x, y, &puyo, &mut out);
                        }
                    }
                }
            }
        }

        write_png(name.as_ref(), &out)
    }

    pub fn export_shape_image<P: AsRef<Path>>(&self, n: usize, name: P) -> io::Result<()> {
        let field = load_sprite("puyos.png")?;
        let puyo = load_sprite("puyos_shape.png")?;
        let mut out = new_canvas(8, 14);
        draw_field(&field, &mut out);

        let shape = self.shapes[n];
        for y in (1..=13).rev() {
            for x in 0..6 {
                if shape.onebit(x, y) > 0 {
                    draw_shape_puyo(&shape, n as u32, x, y, &puyo, &mut out);
                }
            }
        }

        write_png(name.as_ref(), &out)
    }

    fn find_possibilities(&self, colors: &[Color], num: usize) -> Vec<Color> {
        if colors[num] != Color::Empty {
            return vec![colors[num]];
        }
        let mut palette = vec![Color::Red, Color::Blue, Color::Yellow, Color::Green];
        for (index, shape) in self.shapes.iter().enumerate() {
            if index == num {
                continue;
            }
            let expanded = self.shapes[num].expand(*shape);
            if !expanded.is_empty() {
                let vanished = expanded.find_vanishing_bits();
                if !vanished.is_empty() {
                    palette.retain(|color| *color != colors[index]);
                }
            }
        }
        palette
    }

    fn find_color(&self, targets: Vec<Color>, idx: usize) -> Option<BitField> {
        if idx == self.shape_count() {
            let chain_shapes = self.to_chain_shapes_u64_array();
            let overall = self.overall_shape();
            let mut bit_field = BitField::new();
            for x in 0..6 {
                for y in 0..14 {
                    if overall.onebit(x, y) > 0 {
                        for (shape_index, shape) in self.shapes.iter().enumerate() {
                            if shape.onebit(x, y) > 0 {
                                bit_field.set_color(targets[shape_index], x, y);
                                break;
                            }
                        }
                    }
                }
            }

            let mut simulated = bit_field.clone();
            let result = simulated.simulate();
            if result.chains == self.shape_count() {
                let simulated_shapes = bit_field.to_chain_shapes_u64_array();
                let same_chain = simulated_shapes.len() == chain_shapes.len()
                    && simulated_shapes
                        .iter()
                        .zip(chain_shapes.iter())
                        .all(|(left, right)| left == right);
                if same_chain {
                    return Some(bit_field);
                }
            }
            return None;
        }

        for color in self.find_possibilities(&targets, idx) {
            let mut new_targets = targets.clone();
            new_targets[idx] = color;
            if let Some(bit_field) = self.find_color(new_targets, idx + 1) {
                return Some(bit_field);
            }
        }
        None
    }

    fn shift_col(&mut self, insert_shape: FieldBits, shift_target: FieldBits, col_num: usize) {
        for index in 0..self.shapes.len() {
            let mut shape = self.shapes[index];
            let and = shape.and(shift_target);
            let col_bits = and.col_bits(col_num);
            if col_bits == 0 {
                continue;
            }

            let shift_bits = insert_shape.col_bits(col_num).count_ones();
            let shift_col = col_bits << shift_bits;
            if col_num < 4 {
                let shift = FieldBits::with_matrix([shift_col, 0]);
                let mut untarget =
                    FieldBits::with_matrix([!(shift_target.m[0] | (1u64 << (col_num * 16))), 0]);
                untarget.m[0] = untarget.col_bits(col_num);
                untarget = untarget.and(shape);
                match col_num {
                    0 => shape.m[0] &= !0xffff,
                    1 => shape.m[0] &= !0xffff0000,
                    2 => shape.m[0] &= !0xffff00000000,
                    3 => shape.m[0] &= !0xffff000000000000,
                    _ => {}
                }
                shape = shape.or(shift);
                shape = shape.or(untarget);
            } else {
                let shift = FieldBits::with_matrix([0, shift_col]);
                let mut untarget = FieldBits::with_matrix([
                    0,
                    !(shift_target.m[1] | (1u64 << ((col_num - 4) * 16))),
                ]);
                untarget.m[1] = untarget.col_bits(col_num);
                untarget = untarget.and(shape);
                match col_num {
                    4 => shape.m[1] &= !0xffff,
                    5 => shape.m[1] &= !0xffff0000,
                    _ => {}
                }
                shape = shape.or(shift);
                shape = shape.or(untarget);
            }

            self.shapes[index] = shape;
            self.original_shapes[index] = shape;
        }
        self.key_string = None;
    }
}

impl fmt::Display for ShapeBitField {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for y in (1..=14).rev() {
            write!(f, "{y:02}: ")?;
            for x in 0..6 {
                let mut empty = true;
                for (index, shape) in self.shapes.iter().enumerate() {
                    if shape.onebit(x, y) > 0 {
                        write!(f, "{}", num_to_letter(index + 1))?;
                        empty = false;
                        break;
                    }
                }
                if empty {
                    write!(f, ".")?;
                }
            }
            writeln!(f)?;
        }
        Ok(())
    }
}

fn draw_field(puyo: &image::RgbaImage, out: &mut image::RgbaImage) {
    for y in (0..=13).rev() {
        copy_tile(puyo, 5, 0, out, 0, (13 - y) * TILE_SIZE);
        copy_tile(puyo, 5, 0, out, 7 * TILE_SIZE, (13 - y) * TILE_SIZE);
        for x in 0..8 {
            if (x == 0 || x == 7 || y == 0) && y != 13 {
                copy_tile(puyo, 5, 0, out, x * TILE_SIZE, 13 * TILE_SIZE);
            } else {
                copy_tile(puyo, 5, 1, out, x * TILE_SIZE, (13 - y) * TILE_SIZE);
            }
        }
    }
}

fn draw_shape_puyo(
    field_bits: &FieldBits,
    shape_num: u32,
    x: usize,
    y: usize,
    puyo: &image::RgbaImage,
    out: &mut image::RgbaImage,
) {
    let mut tile_y = 0u32;
    let up = field_bits.onebit(x, y + 1) > 0;
    let down = y > 0 && field_bits.onebit(x, y - 1) > 0;
    let left = x > 0 && field_bits.onebit(x - 1, y) > 0;
    let right = x < 5 && field_bits.onebit(x + 1, y) > 0;

    if up && !down && !left && !right {
        tile_y = 1;
    } else if !up && down && !left && !right {
        tile_y = 2;
    } else if up && down && !left && !right {
        tile_y = 3;
    } else if !up && !down && left && !right {
        tile_y = 4;
    } else if up && !down && left && !right {
        tile_y = 5;
    } else if !up && down && left && !right {
        tile_y = 6;
    } else if up && down && left && !right {
        tile_y = 7;
    } else if !up && !down && !left && right {
        tile_y = 8;
    } else if up && !down && !left && right {
        tile_y = 9;
    } else if !up && down && !left && right {
        tile_y = 10;
    } else if up && down && !left && right {
        tile_y = 11;
    } else if !up && !down && left && right {
        tile_y = 12;
    } else if up && !down && left && right {
        tile_y = 13;
    } else if !up && down && left && right {
        tile_y = 14;
    } else if up && down && left && right {
        tile_y = 15;
    }

    overlay_tile(
        puyo,
        shape_num,
        tile_y,
        out,
        ((x + 1) as u32) * TILE_SIZE,
        ((13 - y) as u32) * TILE_SIZE,
    );
}

fn letter_to_num(ch: char) -> Option<usize> {
    SHAPE_LABELS
        .iter()
        .position(|label| *label == ch)
        .map(|idx| idx + 1)
}

fn num_to_letter(num: usize) -> char {
    SHAPE_LABELS
        .get(num - 1)
        .copied()
        .unwrap_or_else(|| panic!("shape label index out of range: {num}"))
}
